pkg/inject/sidecar.go

```diff
@@ -41,21 +41,14 @@ const (
 func Sidecar(jaeger *v1.Jaeger, dep *appsv1.Deployment) *appsv1.Deployment {
 	deployment.NewAgent(jaeger) // we need some initialization from that, but we don't actually need the agent's instance here
 	logFields := jaeger.Logger().WithField("deployment", dep.Name)
-
-	if jaeger == nil || (dep.Annotations[Annotation] != jaeger.Name && dep.Annotations[AnnotationLegacy] != jaeger.Name) {
+	labelValue, hasLabel := dep.Labels[Label]
+	if jaeger == nil || (hasLabel && labelValue != jaeger.Name) {
```

Comment:
Can be the `hasLabel` omitted? e.g.  `if jaeger == nil || (labelValue != jaeger.Name)`

Reply:
I don't think it can be omitted. 

The logic is this: if deployment doesn't have a label, that means that it needs to be injected.  If it has a label, we need to check if the label is equal to jaeger instance. The latest only applies in the case of an update, while the former is for new deployments. Remember, now `inject.Select` function doesn't have side effects, nor modify the deployment.

If we don't check for `hasLabel` some sidecar tests will fail. For instance, `TestInjectSidecar` will fail because it will try to inject a sidecar in a deployment that doesn't have any label (the label is added when the sidecar is injected). If we check only `labelValue != jaeger.Name` the deployment will never be injected. The first time, when it doesn't have a label, `labelValue` will be an empty string.

On the other hand, if we check for `hasLabel` we will inject the side car the first time.  Of course I can do a `labelValue != "",` but I would prefer to use the value returned by go maps for key existence.


Does that makes sense?

Reply:
I'm not a huge fan of this conditional. How about something like this?

```go
	if jaeger == nil {
		logFields.Trace("no Jaeger instance found, skipping sidecar injection")
		return dep
	}

	if val, ok := dep.Labels[Label]; ok && val != jaeger.Name {
		logFields.Trace("deployment is assigned to a different Jager instance, skipping sidecar injection")
		return dep
	}

	decorate(dep)
	logFields.Debug("injecting sidecar")
	dep.Spec.Template.Spec.Containers = append(dep.Spec.Template.Spec.Containers, container(jaeger, dep))
	setInjectedInstanceLabel(dep, jaeger)
```

Reply:
One more thing I wanted to check: this changes the behavior of the `Sidecar` function, so that it won't check the `Annotation` at all anymore. Therefore, this test would be outdated:

https://github.com/jaegertracing/jaeger-operator/blob/4424e8ac6f600a469a2ba9d05abbd895d43a29c4/pkg/inject/sidecar_test.go#L33-L40

Removing the `Annotation` would still get this test passing, which proves the change in behavior. If this is desired, then please change the existing tests, including the one above, but make sure that the contract with the consumers isn't broken: people annotating their deployments have to get the same behavior as before.

Reply:
True, I need to update the above test, (and I need to review if I need to update more). The contract with the consumer isn't broken, the only difference is that now the annotation is only used for select the instance (in the select function) and not for inject. but If I'm not wrong or miss something, that is transparent for the consumer.

Reply:
> I'm not a huge fan of this conditional. How about something like this?
> 
> ```go
> 	if jaeger == nil {
> 		logFields.Trace("no Jaeger instance found, skipping sidecar injection")
> 		return dep
> 	}
> 
> 	if val, ok := dep.Labels[Label]; ok && val != jaeger.Name {
> 		logFields.Trace("deployment is assigned to a different Jager instance, skipping sidecar injection")
> 		return dep
> 	}
> 
> 	decorate(dep)
> 	logFields.Debug("injecting sidecar")
> 	dep.Spec.Template.Spec.Containers = append(dep.Spec.Template.Spec.Containers, container(jaeger, dep))
> 	setInjectedInstanceLabel(dep, jaeger)
> ```

Is essentially the same, but more separated, and with better log messages :)

Reply:
I find it easier to read and understand, but if you disagree, feel free to keep the current version.

Reply:
I indeed think your suggestion is a good idea, improves a lot the readability.